common/network-common/src/main/java/org/apache/spark/network/util/JavaUtils.java

```diff
@@ -362,6 +364,60 @@ public static byte[] bufferToArray(ByteBuffer buffer) {
     }
   }
 
+  /**
+   * Create a temporary directory inside `java.io.tmpdir` with default namePrefix "spark".
+   * The directory will be automatically deleted when the VM shuts down.
+   */
+  public static File createTempDir() throws IOException {
+    return createTempDir(System.getProperty("java.io.tmpdir"), "spark");
+  }
+
+  /**
+   * Create a temporary directory inside the given parent directory. The directory will be
+   * automatically deleted when the VM shuts down.
+   */
+  public static File createTempDir(String root, String namePrefix) throws IOException {
```

Comment:
Is there a particular reason why this is added under src/main and not src/test? 
`createTempDir` could introduce performance issues due to `deleteOnExit` when using the main code.

Reply:
It's used in non-test code. What's the issue with deleteOnExit, or is there a better way to ensure tmp dirs are cleaned up?

Reply:
Sorry, I should have clarified this. DeleteOnExit functionality can cause performance issues due memory leaks under heavy load in certain JDK versions because while the temporary files are deleted, the entries in the map are never cleared.

Reply:
This pr copying the logic of `o.a.spark.util.Utils#createTempDir` to a more basic module due to some underlying modules cannot call `o.a.spark.util.Utils#createTempDir` and in https://github.com/apache/spark/pull/36611 `o.a.spark.util.Utils#createTempDir` also call this method instead.

Initially, `o.a.spark.util.ShutdownHookManager#registerShutdownDeleteDir` was used in `o.a.spark.util.Utils#createTempDir` to clean up temporary files:

https://github.com/apache/spark/blob/12e7991b5b38302e6496307c7263ad729c82a6cf/core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala#L72-L78

But it also cannot be accessed in current module, so here it is changed to use `deleteOnExit`. Personally, there is no essential difference between `registerShutdownDeleteDir` and `deleteOnExit`, if I understand incorrectly, please correct me.


@sadikovi What is the bad case you found? Running UT require more memory or other scenarios? What is your suggestion?

Reply:
Hmm... I know ... `registerShutdownDeleteDir` can be cleanup when call `Utils#deleteRecursively`, but `deleteOnExit` cannot do this ...

Reply:
@srowen @sadikovi 

I have two ways to fix this issue:

1. Revert [SPARK-39204](https://issues.apache.org/jira/browse/SPARK-39204) to break  the call from `o.a.spark.util.Utils#createTempDir` to `JavaUtils#createTempDir`, then all other calls of `JavaUtils#createTempDir()` only exist in test case.

3. https://github.com/apache/spark/pull/40598  try to restore `Utils#createTempDir` use `ShutdownHookManager#registerShutdownDeleteDir` to cleanup `tempDir`

also cc @HyukjinKwon WDYT?

Reply:
Let's just make everything use the explicit shutdown hook mechanism? is there any downside? Like change the JavaUtils 'base' method to just do that.

Reply:
Need move `org.apache.spark.util.ShutdownHookManager` to `network-common` module and rewrite it using Java or make `network-common` module has Scala code

Reply:
If keep using Scala, the work may includes:

1. Move `org.apache.spark.util.ShutdownHookManager` to `org.apache.spark.network.util.ShutdownHookManager` and refactor all other use places.

2. Use `LoggerFactory.getLogger` instead of `extends Logging` due to `Logging` trait in core module

https://github.com/apache/spark/blob/35503a535771d257b517e7ddf2adfaefefd97dad/core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala#L32

3. Need introduce `hadoop-client-api` module to `network-common` module due to  `org.apache.hadoop.util.ShutdownHookManager` used by `SparkShutdownHookManager`, I haven't thought of a solution without introducing `hadoop-client-api` module to `network-common` module

https://github.com/apache/spark/blob/35503a535771d257b517e7ddf2adfaefefd97dad/core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala#L176-L182

Reply:
Maybe it is easier to revert and about the solution in a follow-up PR. Would this work?

Reply:
Oh, OK if that means moving more code around, can we just adjust the Utils.createTempDir implementation? separate it from JavaUtils.createTempDir I guess, which can still be used for test code (or moved to the test tree), or maybe we don't need both

Reply:
I'm not sure about a revert. I'm not sure there is even a problem here.

Reply:
If need to rewrite to Java, will need some additional refactoring work. Are we sure we want to do this? @srowen 

Personally, I think this change is a little large, and if necessary, it is only suitable for Spark 3.5.0？

Reply:
https://github.com/apache/spark/pull/40613/files